### PR TITLE
[engine] fix node seeds for tx with auth node

### PR DESCRIPTION
**Pull request by nickchapman-da**
This fixes the problem with the node-seeds for a TX with an authority node.

The code in `NormalizeRollbacks.scala` post-processes the transaction constructed by `PartialTransaction.scala` to produce a TX which satisfies the rollback-normalization rules. In the process the node-ids are completely rewritten.

As a secondary job, this code returns a list (`seeds`) of (rewritten) node-ids which correspond to a pre-order traversal of the Create and Exercise nodes. This `seeds` list should match up with the `actionNodeSeeds` contained in the original `PartialTransaction`, and is `zip`ed together in `Speedy.finish`

The bug (which is fixed in this PR) is that `NormalizeRollbacks` was calling `pushSeedId` for the `Authority` nodes as it reconstructed the TX. This bug might have been detected if the `zip` in `Speedy.finish` was guarded by an assertion that the two lists/arrays had equal size.

Unfortunately if we add this assertion, it is triggered by the `exer2` test in `RollbackTest.scala` -- I suspect that this indicates a further bug (which I am investigating). ... YES.

Found and fixed bug in `PartialTransaction.endExercises`.

Add the assertion in `Speedy.finish` to ensure the two zipped arrays have the same length,

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala [hunk 1]
```diff
@@ -186,11 +186,9 @@ private[lf] object NormalizeRollbacks {
               }
             }
           case Norm.Aut(aut, subs) =>
-            s.pushSeedId(me) { s =>
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:189]: This one line is wrong, and needs to be removed,

Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala:189]: Along with the matching close curly

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 2]
```diff
@@ -344,11 +344,20 @@ private[lf] object Speedy {
         }
         .to(ImmArray)
 
+    def zipSameLength[X, Y](xs: ImmArray[X], ys: ImmArray[Y]): ImmArray[(X, Y)] = {
+      val n1 = xs.length
+      val n2 = ys.length
+      if (n1 != n2) {
+        sys.error(s"sameLengthZip, $n1 /= $n2")
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:351]: Could you trow an IllegalArgumentException instead ?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 3]
```diff
@@ -344,11 +344,20 @@ private[lf] object Speedy {
         }
         .to(ImmArray)
 
+    def zipSameLength[X, Y](xs: ImmArray[X], ys: ImmArray[Y]): ImmArray[(X, Y)] = {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:347]:
```suggestion
    @throws[IllegalArgumentException]
    def zipSameLength[X, Y](xs: ImmArray[X], ys: ImmArray[Y]): ImmArray[(X, Y)] = {
```

**Review by remyhaemmerle-da — APPROVED** [review]
Good catch!

**Pull request merged**